Mage.Sets/src/mage/cards/u/UnboundFlourishing.java

```diff
@@ -50,38 +47,73 @@ public UnboundFlourishing copy() {
     }
 }
 
-class UnboundFlourishingDoubleXEffect extends ReplacementEffectImpl {
+class UnboundFlourishingDoubleXAbility extends TriggeredAbilityImpl {
 
-    UnboundFlourishingDoubleXEffect() {
-        super(Duration.WhileOnBattlefield, Outcome.Benefit, false);
-        staticText = "Whenever you cast a permanent spell with a mana cost that contains {X}, double the value of X";
+    UnboundFlourishingDoubleXAbility() {
+        super(Zone.BATTLEFIELD, new UnboundFlourishingDoubleXEffect(), false);
+        setTriggerPhrase("Whenever you cast a permanent spell with a mana cost that contains {X}");
     }
 
-    private UnboundFlourishingDoubleXEffect(final UnboundFlourishingDoubleXEffect effect) {
-        super(effect);
+    private UnboundFlourishingDoubleXAbility(final UnboundFlourishingDoubleXAbility ability) {
+        super(ability);
     }
 
     @Override
-    public boolean replaceEvent(GameEvent event, Ability source, Game game) {
-        event.setAmount(CardUtil.overflowMultiply(event.getAmount(), 2));
-        return false;
+    public UnboundFlourishingDoubleXAbility copy() {
+        return new UnboundFlourishingDoubleXAbility(this);
     }
 
     @Override
-    public boolean checksEventType(GameEvent event, Game game) {
-        return event.getType() == GameEvent.EventType.X_MANA_ANNOUNCE;
+    public boolean checkEventType(GameEvent event, Game game) {
+        return event.getType() == GameEvent.EventType.SPELL_CAST;
     }
 
     @Override
-    public boolean applies(GameEvent event, Ability source, Game game) {
-        Spell spell = game.getSpell(event.getTargetId());
-        return spell != null && spell.isPermanent(game) && spell.isControlledBy(source.getControllerId());
+    public boolean checkTrigger(GameEvent event, Game game) {
+        if (!event.getPlayerId().equals(getControllerId())) {
+            return false;
+        }
+        
+        Spell spell = game.getStack().getSpell(event.getTargetId());
+        if (spell != null && spell.isPermanent(game)) {
+            if (spell.getSpellAbility().getManaCostsToPay().containsX()) {
+                game.getState().setValue(this.getSourceId() + UnboundFlourishing.needPrefix, spell);
```

Comment:
Make sure it support multiple triggers/spells. Current code save only one spell per Unbound Flourishing.

Reply:
Good point. I copied it from the copy spell/ability trigger, both of them need to be fixed in that manner.

Reply:
Also don't forget about flourishing ZCC in setValue. So blinked Unbound Flourishing will ignore outdated spell data.

Reply:
A better solution is to not use the global game state, and instead use target pointers or the effect value map.